### ISSUE TITLE
Memory leak while processing AsyncListener annotations

### ISSUE THREAD

**Issue by sbordet** [eclipse/jetty.project#3749]
Class `AsyncListenerServlet.MyAsyncListener` of `test-spec-webapp` is annotated with `@PostConstruct`.

For every request, a new instance of `MyAsyncListener` is created and its annotations processed, but the instance is stored in a list mapped to context attribute `org.eclipse.jetty.lifecyleCallbackCollection` and never removed from the list when the request cycle ends.

```java
AsyncContext asyncContext = req.startAsync();
MyAsyncListener listener = asyncContext.createListener(MyAsyncListener.class);
// listener is never used
asyncContext.complete();
```

The instance is added in `PostConstructAnnotationHandler.doHandle()` which is called directly from `AsyncContext.createListener()`.

Class `PlusDecorator` extracts the attribute and calls `@PostConstruct` for every leaked instance.

I suspect that this happens for field injections too and possibly for other injections as they seems to all use the same mechanism (store instances in a list mapped to a context attribute).

**Comment by sbordet** [comment]
@janbartel I don't understand why we store the instances? 
Would not be enough to create them on the fly, process their annotations (and therefore inject them), and _not_ store them anywhere?

**Comment by janbartel** [comment]
@sbordet  injections and callbacks can be declared either in web.xml or via annotations. In either case, we need to evaluate them, then at the appropriate point in the lifecycle  of the object we need to invoke the callbacks or injections, hence the list. Building the list also helps with quickstart, as we can fully evaluate every callback and injection and save it out to the quickstart-web.xml.

**Comment by janbartel** [comment]
Fixed in 9.4.20 and above.
